server/src/main/java/org/elasticsearch/index/IndexMode.java

```diff
@@ -308,6 +309,87 @@ public SourceFieldMapper.Mode defaultSourceMode() {
         public String getDefaultCodec() {
             return CodecService.BEST_COMPRESSION_CODEC;
         }
+    },
+    LOOKUP("lookup") {
+        @Override
+        void validateWithOtherSettings(Map<Setting<?>, Object> settings) {
+            Integer defaultNumberOfShards = IndexMetadata.INDEX_NUMBER_OF_SHARDS_SETTING.get(Settings.EMPTY);
+            Object providedNumberOfShards = settings.getOrDefault(IndexMetadata.INDEX_NUMBER_OF_SHARDS_SETTING, defaultNumberOfShards);
+            if (Objects.equals(providedNumberOfShards, defaultNumberOfShards) == false) {
+                throw new IllegalArgumentException(
+                    "index with [lookup] mode must have [index.number_of_shards] set to 1 or unset; provided " + providedNumberOfShards
+                );
+            }
+        }
+
+        @Override
+        public Settings defaultIndexSettings() {
+            return Settings.builder()
+                .put(IndexMetadata.SETTING_NUMBER_OF_SHARDS, 1)
+                .put(IndexMetadata.SETTING_AUTO_EXPAND_REPLICAS, "0-all")
```

Comment:
Should we also validate whether auto expand replicas is really enabled? This just defines the defaults, but doesn't enforce that it is actually used?

Reply:
I may be missing something, but the default settings are copied to the index settings when creating an index: https://github.com/elastic/elasticsearch/pull/115143/files#diff-c02a1326ed501a89ce49b7e7a27df4bb9a2256785d00d7278bccb3609334b939R1112

Reply:
I see, the default settings are added after template and create index request settings have been applied. Maybe default settings is the wrong name? Since the settings from index mode don't act as default, but are enforced?

Additionally I think we need to mimic this logic in two other places too for consistency:

1. `MetadataIndexTemplateService#validateIndexTemplateV2(...)` to ensure that index mode default settings are applied during template validation.
2. `TransportSimulateIndexTemplateAction#resolveTemplate(...)` to ensure that index mode default settings are applied during simulate template api.

Reply:
`auto_expand_replicas` setting defaults to `0-all` and can be overridden by users, while the `number_of_shards` setting is enforced and must always be 1.

Reply:
Reading the PR description: `... Lookup indices must have a single shard and be replicated to all data nodes by default. ...`, is just apply a default for `auto_expand_replicas` to `0-all` sufficient?

Reply:
I spoke with Nhat about this and only `number_of_shards` must be 1. Auto expand replicas is optional and seen as an optimization for the enrich join.